RFC-0021-Distributed-Pipeline-Parallel-Technical.md

```diff
@@ -0,0 +1,620 @@
+
+`oneflow.distributed.launch` will launch the processes for each “actor”. Eager mode and graph mode reportedly go down the same (or similar) code path for distributed graph processing, with eager mode using a LazyTensor-like model to capture the program. The compiler knows which rank the compilation is for and will emit code specifically for that rank. Micro-batch splitting and `1f1b` seem to be hard-coded, if they are implemented.
+
+### Pros and Cons of the Approach
+
+**Pro**
+
+* (split pro/con) (**D2**) Not clear if `1f1b` or other schedules are implemented
+* (**D3/D4**) Composable with other parallelism schemes via their “Consistent View” definition
+* (**D5**) `nn.Sequential` not needed, but potentially an `nn.Graph` instance may be needed in some cases
+* (**D6**) Async is probably supportable but not clear. From their presentation, the actor/register model with backpressure can implement on-demand data loading, but I’m not 100% sure what that API looks like
+* (**D7**) Unified programming model that already exists
+
+**Con**
+
+* (split pro/con) (**D2**) Not clear if `1f1b` or other schedules are implemented/implementable?
+* (**D8**) Not clear what the training loop abstraction looks like. The optimizer is installed via an `nn.Graph` API. Loss calculation is created in the `nn.Graph.build()` method.
+
+## Final Analysis
+
+### General Design Axes
+
+Looking through the various approaches above, we can pull out some general design axes we should consider:
+
+
+* (**DA1**) Single-coordinator vs. distributed coordination
+    * Sub-decisions: model execution, autograd, optimizers
+* (**DA2**) Python training loop vs. custom encapsulated training loop (e.g. as in DeepSpeed)
+* (**DA3**) `nn.Sequential` vs. more free-form partitioning
+* (**DA4**) Local vs. distributed loss
+* (**DA5**) Distributed autograd framework vs. manual autograd
+    * As it relates to pipeline schedules
+* (**DA6**) Local vs. distributed optimizer
+* (**DA7**) Synchronous pipeline dispatch vs. asynchronous
+* (**DA8**) Predication scheme (Approach 1 only)
+* (**DA9**) Instruction format (Approach 4 only)
+
+
+We can start analyzing the approaches by these design axes
+
+* Approach 1: SPMD with Predicated Training Loop and message passing
+* Approach 2: RPC with RemoteModule and torchgpipe-style single coordinator (@pritamdamania87 RFC)
+    * Single-coordinator
+        * CUDA command buffer analogy
+    * Continuation passing
+* Approach 3: MPMD with RemoteModule and message passing (fairscale experimental)
+* Approach 4: MPMD with a custom interpreter/instruction format and message passing (DeepSpeed)
+* Approach 5: RPC with remote modules and generalized Module-server architecture (SageMaker)
+* Approach 6: SPMD with Program capture/JIT compilation and message passing (OneFlow)
```

Comment:
I am curious how quickly we can prototype some of the approaches on typical transformer models such as GPT3-13B or GPT3-175B, and then evaluate them on both UX and performance aspects.

IMO from pure UX point of view we might want to preserve **Python training loop** and **free-form (no sequential) authoring**, otherwise the adoption barrier might be so high that users would just not try it.

I even feel that maximizing UX at a small cost of perf is fine, because user can always find a way (and hire help) to hyper-optimize their training pipeline later if needed.

Reply:
Thanks @yf225! Indeed, we are also hearing feedback from other users that the front-end restrictions will be prohibitive. I think this week I will likely keep iterating on both strategy (prioritize non-`Sequential` models) and technical approaches that address the front-end issue and update this RFC.